File: cosmos/src/main/java/co/sentinel/cosmos/network/ChannelBuilder.java
package co.sentinel.cosmos.network;

import java.util.concurrent.Executors;

import co.sentinel.cosmos.BuildConfig;
import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;

public class ChannelBuilder {

    private static String grpcSentinelMain = BuildConfig.GRPC_BASE_URL;
    private static int portSentinelMain = BuildConfig.GRPC_PORT;


    public final static int TIME_OUT = 30;

    //Channel for sentinel main
    private static ManagedChannel channel_sentinel_main = null;

    public static ManagedChannel getMainChannel() {
        if (channel_sentinel_main == null) {
            synchronized (ChannelBuilder.class) {
                channel_sentinel_main = ManagedChannelBuilder
                        .forAddress(grpcSentinelMain, portSentinelMain)
                        .usePlaintext()
                        .executor(Executors.newSingleThreadExecutor())
                        .build();
            }
        }
        return channel_sentinel_main;
    }

    public static void resetSentinelMain() {
        if (channel_sentinel_main != null) {
            synchronized (ChannelBuilder.class) {
                channel_sentinel_main.shutdownNow();
                channel_sentinel_main = ManagedChannelBuilder
                        .forAddress(grpcSentinelMain, portSentinelMain)
                        .usePlaintext()
                        .executor(Executors.newSingleThreadExecutor())
                        .build();
            }
        }
    }

    public static void createCustomChannel(String address, int port) {
        synchronized (ChannelBuilder.class) {
            if (channel_sentinel_main != null) {
                channel_sentinel_main.shutdownNow();
            }
            grpcSentinelMain = address;
            portSentinelMain = port;
            channel_sentinel_main = ManagedChannelBuilder
                    .forAddress(grpcSentinelMain, portSentinelMain)
                    .usePlaintext()
                    .executor(Executors.newSingleThreadExecutor())
                    .build();
        }
    }
}

File: cosmos/src/main/java/co/sentinel/cosmos/network/ChannelBuilder.kt
package co.sentinel.cosmos.network

import android.content.Context
import co.sentinel.cosmos.BuildConfig
import io.grpc.ManagedChannel
import io.grpc.ManagedChannelBuilder
import java.util.concurrent.Executors
import timber.log.Timber

class ChannelBuilder(private val context: Context) {

  @Volatile
  private var mainChannel: ManagedChannel? = null

  private val prefs by lazy { context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE) }

  fun getMainChannel(): ManagedChannel {
    mainChannel?.let { channel -> return channel }
    return synchronized(this) {
      mainChannel?.let { channel -> return@synchronized channel }
      val address = prefs.getString(PREFS_ADDRESS, null) ?: BuildConfig.GRPC_BASE_URL
      val port = prefs.getInt(PREFS_PORT, BuildConfig.GRPC_PORT)
      Timber.d("Cosmos channel created with the endpoint: $address:$port")
      ManagedChannelBuilder
        .forAddress(address, port)
        .usePlaintext()
        .executor(Executors.newSingleThreadExecutor())
        .build()
    }
      .also { mainChannel = it }
  }

  fun setCustomEndpoint(address: String, port: Int) {
    synchronized(this) {
      if (this.mainChannel != null) {
        this.mainChannel!!.shutdownNow()
        this.mainChannel = null
      }
      prefs.edit()
        .putString(PREFS_ADDRESS, address)
        .putInt(PREFS_PORT, port)
        .apply()
      Timber.d("Cosmos endpoint has been changed: $address:$port")
    }
  }

  fun getCurrentEndpoint(): Pair<String, Int> {
    val address = prefs.getString(PREFS_ADDRESS, null) ?: BuildConfig.GRPC_BASE_URL
    val port = prefs.getInt(PREFS_PORT, BuildConfig.GRPC_PORT)
    return address to port
  }

  companion object {
    private const val PREFS_NAME = "CosmosChannelPrefs"
    private const val PREFS_ADDRESS = "PrefsAddress"
    private const val PREFS_PORT = "PrefsPort"

    const val TIME_OUT = 30L
    const val TIME_OUT_BROADCAST = 60L
  }
}
